train.py

```diff
@@ -6,51 +8,300 @@
+                self.best_state = best_state_dict['state']
+                self.lowest_val_loss = best_state_dict['loss']
+        else:
+            self._loaded_metrics = False
+
+    @functools.partial(jax.jit, static_argnums=0)
+    def _evaluate_step(
+            self, state: dict, graphs: jraph.GraphsTuple) -> float:
+        """Calculate the mean loss for a batch of graphs."""
+        mean_loss = self._loss_fn(state['params'], graphs, self._net_apply)
+        return mean_loss
+
+    def evaluate_split(
+            self, 
+            state: dict,
+            graphs: Sequence[jraph.GraphsTuple],
+            batch_size: int) -> float:
+        """Return mean loss for all graphs in graphs."""
+        
+        reader = DataReader(data=graphs, 
+            batch_size=batch_size, repeat=False)
+        
+        loss_list = [self._evaluate_step(state, batch) for batch in reader]
+        return np.mean(loss_list)
+
+    def evaluate_model(
+            self, 
+            state: Dict,
+            datasets: Dict[str, Iterable[jraph.GraphsTuple]],
+            splits: Iterable[str]) -> Dict[str, float]:
+        """Return mean loss for every split in splits."""
+        eval_loss = {}
+        for split in splits:
+            eval_loss[split] = self.evaluate_split(state, 
+                    datasets[split].data, datasets[split].batch_size)
+            if split=='validation':
+                if self.best_state is None or eval_loss[split] < self.lowest_val_loss:
+                    self.best_state = state.copy()
+                    self.lowest_val_loss = eval_loss[split]
+        return eval_loss
+
+    def init_loss_lists(self, splits):
+        """Initialize a dict to save evaluation losses in."""
+        self.loss_dict = {}
+        if not self._loaded_metrics:
+            for split in splits:
+                self.loss_dict[split] = []
+            # initialize a queue with validation losses for early stopping
+            self.early_stopping_queue = []
+        else:
+            for split in splits:
+                self.loss_dict[split] = self._metrics_dict[split]
+            self.early_stopping_queue = self._metrics_dict['queue']
+    
+    def save_losses(self, loss_dict, splits, step):
+        """Append values in loss_dict to the object values in self.loss_dict for all splits.
+        Also create the local early stopping queue."""
+        for split in splits:
+            self.loss_dict[split].append([step, loss_dict[split]])
+            if split == 'validation':
+                self.early_stopping_queue.append(loss_dict[split])
+
+    def check_early_stopping(self):
+        """Check the early stopping criterion. If the newest validaiton loss in 
+        self.early_stopping_queue is higher than the zeroth one return True for early stopping.
+        Otherwise, delete the zeroth element in queue and return False for no early stopping."""
+        queue = self.early_stopping_queue # abbreviation
```

Comment:
For inline comments like this one you need two spaces before the inline comment.

Ex.

```
        queue = self.early_stopping_queue  # Abbreviation.
```